### PR TITLE
Major revisions

### LINKED ISSUE #27: name unicode is not defined
This line: 

https://github.com/RadioAstronomySoftwareGroup/healvis/blob/a7abbfb948d0471e112a368ed37926496a4de077/healvis/sky_model.py#L42

is not python-3 compatible, and prevents even the installation of this package on python 3 (since this module is imported in the ``setup.py`` via the importing of the entire package to get ``version``).

### PR DESCRIPTION
**Pull request by aelanman**
Fixes a bunch of issues, listed in the change log. The biggest change is that the alt/az calculation is fixed such that it matches the astropy calculation to a respectable precision. Altitudes (or, rather, zenith angles) match to 1e-5, and did before these changes. The azimuths, however, were off due to an erroneous assumption about the North pole position in ICRS (not accounting for the drift of the Earth's north pole with respect to the north celestial pole). They are now correct to within an arcminute, when previous they were off by a half degree in some cases.

For this big a change, I've raised the version number, since data generated before and after this _may_ differ slightly.

Closes #27

### CODE REVIEW

**Comment by aelanman** [comment]
I've tagged @steven-murray to review because this addresses an issue he raised.

File: healvis/observatory.py [hunk 1]
```diff
@@ -113,40 +113,66 @@ def set_pointings(self, time_arr):
         """
         Set the pointing centers (in ra/dec) based on array location and times.
             Dec = self.lat
+        Also sets the north pole positions in ICRS.
         RA  = What RA is at zenith at a given JD?
         """
         telescope_location = EarthLocation.from_geodetic(self.lon * units.degree, self.lat * units.degree)
+        self.telescope_location = telescope_location
```
Comment (steven-murray) [inline healvis/observatory.py]: This happens a bit throughout `healvis`, and I don't want to suggest to change everything, but I thought I would mention that I find it really hard to follow code that sets instance attributes outside of `__init__`

Comment (aelanman) [reply healvis/observatory.py]: Good point. I've moved that particular assignment to the init function, and added some more documentation on other attributes, even if they're not set there.

File: healvis/observatory.py [hunk 2]
```diff
@@ -219,9 +249,19 @@ def get_observed_region(self, Nside):
     def vis_calc(self, pcents, tinds, shell, vis_array, Nfin, beam_pol='pI'):
         if len(pcents) == 0:
             return
+
+        haspoles = True
+        if not hasattr(self, 'north_poles'):
```
Comment (steven-murray) [inline healvis/observatory.py]: how would it be possible that the instance has no attribute `north_poles`? Is it possible to _not_ call `get_uvw`?

File: healvis/observatory.py [hunk 3]
```diff
@@ -219,9 +249,19 @@ def get_observed_region(self, Nside):
     def vis_calc(self, pcents, tinds, shell, vis_array, Nfin, beam_pol='pI'):
```
Comment (steven-murray) [inline healvis/observatory.py]: need a docstring!

Comment (steven-murray) [reply healvis/observatory.py]: possibly also more descriptive variable names? `pcents` I get means pointing centres, though that's not _immediately_ clear. But `tinds` I hae no idea, and `Nfin` either.

File: healvis/observatory.py [hunk 4]
```diff
@@ -261,6 +300,9 @@ def make_visibilities(self, shell, Nprocs=1, beam_pol='pI'):
         conv_fact = jy2Tsr(np.array(self.freqs), bm=pix_area_sr)
         self.Ntimes = len(self.pointing_centers)
 
+        if self.pointing_centers is None:
```
Comment (steven-murray) [inline healvis/observatory.py]: would it be possible for this method just to call `set_pointings()` itself?

File: healvis/observatory.py [hunk 5]
```diff
@@ -269,7 +311,7 @@ def make_visibilities(self, shell, Nprocs=1, beam_pol='pI'):
         Nfin = mp.Value('i', 0)
 
         if Nprocs > 1 and not isinstance(shell.data, mparray):
-            warnings.warn("Caution: SkyModel data array is not in shared memory. With Nprocs > 1, this will cause duplication")
+            warnings.warn("Caution: SkyModel data array is not in shared memory. With Nprocs > 1, this will cause duplication.")
```
Comment (steven-murray) [inline healvis/observatory.py]: is there an easy way to make it in shared memory? Is that documented somewhere?

File: healvis/simulator.py [hunk 6]
```diff
@@ -175,6 +175,12 @@ def parse_frequency_params(freq_params):
         # Check that the freq_array is consistent with channel_width
         assert np.allclose(np.diff(freq_arr), np.ones(Nfreqs - 1) * channel_width)
 
+    if 'freq_chans' in freq_params:
+        chans = ast.literal_eval(freq_params['freq_chans'])
+        freq_arr = freq_arr[:, slice(*chans)]
+        Nfreqs = freq_arr.size
+    print("Freq array:" + str(freq_arr.shape))
```
Comment (steven-murray) [inline healvis/simulator.py]: this has the smell of a debug comment

File: healvis/simulator.py [hunk 7]
```diff
@@ -455,6 +469,7 @@ def setup_uvdata(array_layout=None, telescope_location=None, telescope_name=None
     for (a1, a2) in bls:
         bl_array.append(uvutils.antnums_to_baseline(a1, a2, 1))
     bl_array = np.asarray(bl_array)
+    print("Nbls: {}".format(bl_array.size))
```
Comment (steven-murray) [inline healvis/simulator.py]: again, smells like a debug print. If so, but you want to keep it around, you should possibly convert it to a logging call.

File: healvis/sky_model.py [hunk 8]
```diff
@@ -175,8 +176,9 @@ def read_hdf5(self, filename, freq_chans=None, shared_memory=False, do_not_overw
                 sky_freqs_full = infile['freqs'][()]
                 freq_chans = np.where(np.in1d(sky_freqs_full, self.freqs))[0]
                 sky_freqs_part = sky_freqs_full[freq_chans]
-                if not np.allclose(self.freqs, sky_freqs_part):
-                    raise ValueError("Currently set frequencies do not match any subset of file's frequencies.")
+                if self.freqs is not None:
```
Comment (steven-murray) [inline healvis/sky_model.py]: second "if" could be an "and" on the first one

File: healvis/tests/test_observatory.py [hunk 9]
```diff
@@ -240,3 +241,42 @@ def test_gsm_pointing():
 
     # make sure peak is at time index 5, centered at transit time
     nt.assert_equal(np.argmax(np.abs(visibs[:, 0, 0])), 5)
+
+
+def test_az_za_astropy():
+    """
+    Check the calculated azimuth and zenith angle for a selection of HEALPix pixels against the corresponding astropy calculation.
```
Comment (steven-murray) [inline healvis/tests/test_observatory.py]: line length

File: healvis/tests/test_observatory.py [hunk 10]
```diff
@@ -240,3 +241,42 @@ def test_gsm_pointing():
...
+    obs = observatory.Observatory(latitude, longitude)
+
+    t0 = Time(2458684.453187554, format='jd')
+    obs.set_fov(180)
+
+    zen = AltAz(alt=Angle('90d'), az=Angle('0d'), obstime=t0, location=loc)
+
+    zen_radec = zen.transform_to(ICRS)
+    center = [zen_radec.ra.deg, zen_radec.dec.deg]
+    north_radec = AltAz(alt='90.0d', az='0.0d', obstime=t0, location=EarthLocation.from_geodetic(lat='90.d', lon='0d', height=0.0)).transform_to(ICRS)
```
Comment (steven-murray) [inline healvis/tests/test_observatory.py]: line length

File: healvis/tests/test_observatory.py [hunk 11]
```diff
@@ -240,3 +241,42 @@ def test_gsm_pointing():
...
+    yvec = np.array([north_radec.ra.deg, north_radec.dec.deg])
+    obs.telescope_location = loc
+    za, az, inds = obs.calc_azza(Nside, center, yvec, return_inds=True)
+
+    ra, dec = hp.pix2ang(Nside, inds, lonlat=True)
+
+    altaz_astropy = ICRS(ra=Angle(ra, unit='deg'), dec=Angle(dec, unit='deg')).transform_to(AltAz(obstime=t0, location=loc))
```
Comment (steven-murray) [inline healvis/tests/test_observatory.py]: line length

File: healvis/observatory.py [hunk 12]
```diff
@@ -113,40 +113,66 @@ def set_pointings(self, time_arr):
...
 
-    def calc_azza(self, Nside, center, return_inds=False):
+    def calc_azza(self, Nside, center, north=None, return_inds=False):
         """
-        Set the az/za arrays.
+
+        Calculate azimuth/altitude of sources given the pointing center.
+
+        Parameters:
             Center = lon/lat in degrees
```
[the remark this answers is not in the document]
Comment (aelanman) [reply healvis/observatory.py]: Definitely another PR.

File: healvis/observatory.py (hunk 3 above)
Comment (aelanman) [reply healvis/observatory.py]: This function is run internally, sent to subprocesses via the `multiprocessing` module. I've made it a private function and added some documentation.

File: healvis/observatory.py (hunk 4 above)
Comment (aelanman) [reply healvis/observatory.py]: Only if it is given a set of times. I've added that as an optional keyword and allowed this method to call `set_pointings`.

File: healvis/observatory.py (hunk 5 above)
Comment (aelanman) [reply healvis/observatory.py]: It's not well documented, but the shared memory class is in utils.py as `mparray`. Reading in the sky model using `SkyModel.read_hdf5`, there is an option to read the data into a shared memory array. This is the default behavior of `simulator.run_simulation`.

File: healvis/simulator.py (hunk 7 above)
Comment (aelanman) [reply healvis/simulator.py]: It is. I've gotten rid of it.

File: healvis/simulator.py (hunk 6 above)
Comment (aelanman) [reply healvis/simulator.py]: This was removed in a recent commit.

File: healvis/sky_model.py (hunk 8 above)
Comment (aelanman) [reply healvis/sky_model.py]: Fixed. Thanks!

File: healvis/observatory.py (hunk 2 above)
Comment (aelanman) [reply healvis/observatory.py]: In a recent change (initializing all instance variables in the __init__) the north_poles attribute is set to None, so I changed this to check if it isn't None before proceeding.

**Comment by aelanman** [comment]
Hi Steven. Thanks for the feedback! I've tried to address most of your comments. I should note that this PR is focused on fixing specific issues with the az/za calculation and passing frequency interpolation settings. Most style and documentation issues will be left for future work.

**Comment by steven-murray** [comment]
Looks like you need a review from @nkern as well, since I don't have write access.

File: healvis/observatory.py [hunk 13]
```diff
@@ -230,14 +283,13 @@ def vis_calc(self, pcents, tinds, shell, vis_array, Nfin, beam_pol='pI'):
...
-                #        print('Mem: {}GB'.format(memory_usage_GB))
-                #        sys.stdout.flush()
                 if Nfin.value > 0:
+                    dt = (time.time() - self.time0)
                     sys.stdout.write('Finished: {:d}, Elapsed {:.2f}min, Remain {:.3f}hour, MaxRSS {}GB\n'.format(
                         Nfin.value, dt / 60., (1 / 3600.) * (dt / float(Nfin.value)) * (self.Ntimes - Nfin.value), memory_usage_GB))
                     sys.stdout.flush()
 
-    def make_visibilities(self, shell, Nprocs=1, beam_pol='pI'):
+    def make_visibilities(self, shell, Nprocs=1, times_jd=None, beam_pol='pI'):
```
Comment (nkern) [inline healvis/observatory.py]: if `vis_calc` is now a hidden function, this probably needs a more fleshed-out doc-string with input parameter types etc.
